src/ngraph/op/non_zero.cpp

```diff
@@ -70,3 +74,98 @@ shared_ptr<Node> op::v3::NonZero::clone_with_new_inputs(const OutputVector& new_
     check_new_args_count(this, new_args);
     return make_shared<v3::NonZero>(new_args.at(0), m_output_type);
 }
+
+namespace
+{
+    template <element::Type_t INPUT_ET, element::Type_t OUT_ET>
+    bool evaluate_nonzero_execute(const HostTensorPtr& input, const HostTensorPtr& output)
+    {
+        using T = typename element_type_traits<INPUT_ET>::value_type;
+        using U = typename element_type_traits<OUT_ET>::value_type;
+
+        Shape input_shape = input->get_shape();
+        size_t input_rank = input_shape.size();
+
+        size_t non_zero_count =
+            runtime::reference::non_zero_get_count<T>(input->get_data_ptr<INPUT_ET>(), input_shape);
+
+        Shape out_shape;
+        if (non_zero_count == 0)
+        {
+            out_shape = Shape{0};
+        }
+        else if (input_rank == 0)
+        {
+            out_shape = Shape{1, 1};
+        }
+        else
+        {
+            out_shape = Shape{input_rank, non_zero_count};
+        }
+
+        output->set_shape(out_shape);
+        runtime::reference::non_zero<T, U>(
+            input->get_data_ptr<INPUT_ET>(), output->get_data_ptr<OUT_ET>(), input_shape);
+
+        return true;
+    }
+
+    template <element::Type_t INPUT_ET>
+    bool evaluate(const HostTensorPtr& input, const HostTensorPtr& output)
+    {
+        if (output->get_element_type() == element::i64)
+        {
+            return evaluate_nonzero_execute<INPUT_ET, element::Type_t::i64>(input, output);
+        }
+        else
+        {
+            return evaluate_nonzero_execute<INPUT_ET, element::Type_t::i32>(input, output);
+        }
+    }
+
+    bool evaluate_nonzero(const HostTensorPtr& input, const HostTensorPtr& output)
+    {
+        bool rc = true;
+
+        switch (input->get_element_type())
+        {
+            TYPE_CASE(i8)(input, output);
+            break;
+            TYPE_CASE(i16)(input, output);
+            break;
+            TYPE_CASE(i32)(input, output);
+            break;
+            TYPE_CASE(i64)(input, output);
+            break;
+            TYPE_CASE(u8)(input, output);
+            break;
+            TYPE_CASE(u16)(input, output);
+            break;
+            TYPE_CASE(u32)(input, output);
+            break;
+            TYPE_CASE(u64)(input, output);
+            break;
+            TYPE_CASE(bf16)(input, output);
+            break;
+            TYPE_CASE(f32)(input, output);
+            break;
+            TYPE_CASE(f64)(input, output);
+            break;
+        default: rc = false; break;
+        }
+        return rc;
+    }
+}
+
+bool op::v3::NonZero::evaluate(const HostTensorVector& outputs, const HostTensorVector& inputs)
+{
+    if (outputs[0]->get_element_type() != element::i64 &&
```

Comment:
Why don't we do this one as a switch with default returning `false` so it's symmetric with the inputs selection switches, and it will be simpler to add new types later.

Reply:
Done